### ISSUE TITLE
Support atomic operations on update

### ISSUE THREAD

**Issue by thebiglabasky** [prisma/prisma-client-js#775]
## Problem

Whenever in need to perform an update on a field based on its current value, Prisma requires to first fetch the record, and then perform the update call.
This creates room for race conditions if the value gets changed in the data source between the fetching and updating calls.
The alternative is to rely on raw SQL queries, which defeats the value of using Prisma.

## Context
SQL offers to perform a certain set of operations on fields in an UPDATE query:
 * Arithmetic operations: add (`+`), subtract (`-`), multiply (`*`), divide (`/`) or modulo (`%`)
 * String operations: concat (`field = otherField + ' '`)
 * Functions (type-dependent): e.g. `CONCAT`, `LOWER`, `UPPER` for strings, `ABS`, `CEIL`, `ROUND`, etc. for numbers

## Ideal solution
An ideal solution would offer a way to express operations, and allow to incrementally add support for more and more functions over time, while only supporting a limited set in the first iteration.

## Proposed first iteration
The most important set of operations to support are the *arithmetic operations on numbers*, provided that they are the most commonly used.
Type-dependent functions supports could be considered separately.

## Suggested solutions
 * cf. #655 for a set of proposals

**Comment by phillip055** [comment]
This would be great with transaction api

**Comment by matthewmueller** [comment]
**Suggested Solution**

We go with the **Object** notation, unless it's easier to go straight to the **Helper Function** syntax.

**Process**

The following was my process for working through the solution:

1. I started with S-expressions since they are the most "pure" expression-syntax.
2. I turned those S-expressions into Javascript via functions.
3. Those functions morphed into the AST representation.
4. I simplified the AST into a more domain-specific object notation.
5. Build up helper functions from the AST.

<details>
<summary><b>1. S-expressions</b></summary>

Initial exploration using S-expressions. Here, everything is a list.

```clojure
age: (+ age 10)
age: (- age 10)
age: (* age 10)
age: (/ age 10)
age: (% age 10)

age: (+ age (+ otherField 10))
age: (- age (- otherField 10))
age: (* age (* otherField 10))
age: (/ age (/ otherField 10))
age: (% age (% otherField 10))

full_name: (concat first_name last_name)
checksum: (md5 contents)
contents_b64: (base64 contents)
email: (lower email)
amount: (abs amount)
amount: (ceil 10.5)
amount: (round 10.5)
```

A direct translation, doesn't lend itself very well to Javascript `["+", Symbol("age"), ["+", Symbol("otherField"), 10]]`.

</details>

<details>
<summary><b>2. Functions</b></summary>

This form uses functions:

```js
age: add(age, 10)
age: subtract(age, 10)
age: multiply(age, 10)
age: divide(age, 10)
age: modulo(age, 10)

age: add(age, add(otherField, 10))
age: subtract(age, subtract(otherField, 10))
age: multiply(age, multiply(otherField, 10))
age: divide(age, divide(otherField, 10))
age: modulo(age, modulo(otherField, 10))

full_name: concat(first_name, last_name)
checksum: md5(contents)
contents_b64: base64(contents)
email: lower(email)
amount: abs(amount)
amount: ceil(10.5)
amount: round(10.5)
```

I feel like this would be the ideal solution if these functions were magically global.

</details>

<details>
<summary><b>3. AST</b></summary>

Using https://astexplorer.net/, we could turn these functions into proper AST objects:

```js
age: {
  type: "CallExpression",
  callee: {
    type: "Identifier",
    name: "add"
  },
  arguments: [
    {
      type: "Identifier",
      name: "age"
    },
    {
      type: "Literal",
      value: 10
    },
  ]
}

age: {
  type: "CallExpression",
  callee: {
    type: "Identifier",
    name: "add"
  },
  arguments: [
    {
      type: "Identifier",
      name: "age"
    },
    {
      type: "CallExpression",
      callee: {
        type: "Identifier",
        name: "add"
      },
      arguments: [
        {
          type: "Identifier",
          name: "otherField"
        },
        {
          type: "Literal",
          value: 10
        }
      ]
    }
  ]
}


age: {
  type: "CallExpression",
  callee: {
    type: "Identifier",
    name: "round"
  },
  arguments: [
    {
      type: "Literal",
      value: 10
    }
  ]
}
```

This would work, but it'd be a bit verbose for people to write. It's also somewhat unnecessary, since Rust is eventually the one executing these expressions and Rust doesn't have an understanding of the Javascript AST.

</details>

**4. Objects**

Since we don't need the full expressiveness of the Javascript AST, we can simplify our AST:

```js
age: { fn: "add", args: [ Symbol("age"), 10 ] }
age: { fn: "subtract", args: [ Symbol("age"), 10 ] }
age: { fn: "multiply", args: [ Symbol("age"), 10 ] }
age: { fn: "divide", args: [ Symbol("age"), 10 ] }
age: { fn: "modulo", args: [ Symbol("age"), 10 ] }

age: { fn: "add", args: [ Symbol("age"), { fn: "add", args: [ Symbol("otherField"), 10 ] } ] }
age: { fn: "subtract", args: [ Symbol("age"), { fn: "subtract", args: [ Symbol("otherField"), 10 ] } ] }
age: { fn: "multiply", args: [ Symbol("age"), { fn: "multiply", args: [ Symbol("otherField"), 10 ] } ] }
age: { fn: "divide", args: [ Symbol("age"), { fn: "divide", args: [ Symbol("otherField"), 10 ] } ] }
age: { fn: "modulo", args: [ Symbol("age"), { fn: "modulo", args: [ Symbol("otherField"), 10 ] } ] }

full_name: { op: "concat", args: [ Symbol("first_name"), Symbol("last_name") ] }
checksum: { op: "md5", args: [ Symbol("contents") ] }
amount: { op: "round", args: [ 10.5 ] }
```

A bit verbose, but it does give us the flexibility to add more operations and will allow us to introduce method expressions later.

Note on Symbols: are like creating your own custom identifier. I'm not sure if we should use them here, but the application does feel appropriate. I'd imagine these would be imported from Prisma Client. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol.

- Open Questions:
  - Should we use symbols?
  - Can we make this type-safe? Filters would depend on the column type,
    - e.g. `age` supports number operations
    - e.g. `full_name` supports string operations

**5. Helper functions**

We could eventually do this later.

```js
age: (x) => x.add(x.age, 10)
age: (x) => x.subtract(x.age, 10)
age: (x) => x.multiply(x.age, 10)
age: (x) => x.divide(x.age, 10)
age: (x) => x.modulo(x.age, 10)

age: (x) => x.add(x.age, x.add(x.otherField, 10))
age: (x) => x.subtract(x.age, x.subtract(x.otherField, 10))
age: (x) => x.multiply(x.age, x.multiply(x.otherField, 10))
age: (x) => x.divide(x.age, x.divide(x.otherField, 10))
age: (x) => x.modulo(x.age, x.modulo(x.otherField, 10))

age: (x) => x.concat(x.first_name, x.last_name)
checksum: (x) => x.md5(x.contents)
amount: (x) => x.round(10.5)
```

- Alternatives
  - Use a separate argument for fields
  - `x.age.add(10)` form instead

**Comment by matthewmueller** [comment]
After some additional internal discussion, we're leaning towards starting with a simpler API and introduce expressions later.

**Step 1: Limited object syntax**

Start with something simple. A limited solution to atomic updates.

```js
age: {
    set: 18, // this just sets the value, previous syntax is `data: { age: 18 }`
    increment: 1, // age = age + 1
    decrement: 2, // age = age - 2
    multiply: 3, // age = age * 3
    divide: 4, // age = age / 4
}
```

**Step 2: Add functional syntax**

The aim here is to provide higher-level syntax-sugar. The underlying query that this generates may be equivalent to the object expression above, but it doesn't have to be. It could be a more generic expression syntax.

```js
{ 
   age: u => u.age.inc(1)
}
```

**Step 3: Extend Functional Syntax**

```js
{
  age: u => u.age.add(u.otherField.add(10))
}
```

We'll later extend this syntax to support more arbitrary expressions. The underlying query that this generates will likely need to be more generic than the object syntax above to accommodate this.

Right now we're planning on starting with **Step 1**. **Step 2** and **Step 3** are currently unplanned and the syntax may change.

**Open Questions**

1. It might be nice to add a couple simple string operations to stress test our solution. I think `concat` would be the most useful, but that also brings complexity. Two other options could be `append` and `prepend`
2. Are there any other must-have filters in **Step 1** that we're missing?
3. What will the protocol look like between client and query engine?

**Comment by thebiglabasky** [comment]
We released the Step 1 of the solution in [release 2.6.0](https://github.com/prisma/prisma/releases/tag/2.6.0).

We will keep tabs on [your feedback](https://github.com/prisma/prisma/issues/3108) and possibly follow up with an expression-based notation in a separate issue.
Thanks!
